mixnet/sim/simulation.py

```diff
@@ -0,0 +1,121 @@
+import usim
+from matplotlib import pyplot
+
+import framework.usim as usimfw
+from framework import Framework
+from protocol.config import GlobalConfig, MixMembership, NodeInfo
+from protocol.node import Node, PeeringDegreeReached
+from sim.config import Config
+from sim.connection import (
+    MeteredRemoteSimplexConnection,
+    ObservedMeteredRemoteSimplexConnection,
+)
+from sim.state import NodeState, NodeStateTable
+from sim.stats import ConnectionStats
+
+
+class Simulation:
+    """
+    Manages the entire cycle of simulation: initialization, running, and analysis.
+    """
+
+    def __init__(self, config: Config):
+        self.config = config
+
+    async def run(self):
+        # Run the simulation
+        conn_stats, node_state_table = await self.__run()
+        # Analyze the simulation results
+        conn_stats.analyze()
+        node_state_table.analyze()
+        # Show plots
+        if self.config.simulation.show_plots:
+            pyplot.show()
+
+    async def __run(self) -> tuple[ConnectionStats, NodeStateTable]:
+        # Initialize analysis tools
+        node_state_table = NodeStateTable(
+            self.config.network.num_nodes, self.config.simulation.duration_sec
+        )
+        conn_stats = ConnectionStats()
+
+        # Create a μSim scope and run the simulation
+        async with usim.until(usim.time + self.config.simulation.duration_sec) as scope:
+            self.framework = usimfw.Framework(scope)
+            nodes = self.__init_nodes(node_state_table, conn_stats)
+            for node in nodes:
+                self.framework.spawn(self.__run_node_logic(node))
+
+        # Return analysis tools once the μSim scope is done
+        return conn_stats, node_state_table
+
+    def __init_nodes(
+        self, node_state_table: NodeStateTable, conn_stats: ConnectionStats
+    ) -> list[Node]:
+        # Initialize node/global configurations
+        node_configs = self.config.node_configs()
+        global_config = GlobalConfig(
+            MixMembership(
+                [
+                    NodeInfo(node_config.private_key.public_key())
+                    for node_config in node_configs
+                ],
+                self.config.mix.mix_path.seed,
+            ),
+            self.config.mix.transmission_rate_per_sec,
+            self.config.mix.max_message_size,
+            self.config.mix.mix_path.max_length,
+        )
+
+        # Initialize Node instances
+        nodes = [
+            Node(self.framework, node_config, global_config)
+            for node_config in node_configs
+        ]
+
+        # Connect nodes to each other
+        meter_start_time = self.framework.now()
+        for i, node in enumerate(nodes):
+            # For now, we only consider a simple ring topology for simplicity.
```

Comment:
For normal simulations, we cannot use ring topology. We should rather randomly connect nodes based on their peering degree.

Reply:
Yes, exactly. As I left a comment https://github.com/logos-co/nomos-simulations/pull/6#discussion_r1683899827, I will open a new PR for building a topology randomly (but also deterministically). For now, the ring topology is good to check whether the simulation works well. For example, we can see that all nodes produce the same amount of bandwidth usage because they have the same number of connections. But yes, we have to use the random topology for realistic simulation.

Reply:
I added a TODO comment: f062837 to not forget. I'll open a PR for this.

Reply:
I just implemented the random topology generation, and opened a PR: https://github.com/logos-co/nomos-simulations/pull/10